### PR TITLE
special case complex division for zero imaginary part

### LINKED ISSUE #48457: Change in ComplexF64 division
https://github.com/JuliaLang/julia/pull/47255 caused the following:

```julia
julia> inv(0.2 + 0.0im)
4.999999999999999 - 0.0im
```

In 1.8 this was:

```julia
julia> inv(0.2 + 0.0im)
5.0 - 0.0im
```

I don't know what the guarantees are for this operation but it seems kind of silly to stop being able to do this.

### PR DESCRIPTION
**Pull request by KristofferC**
Fixes https://github.com/JuliaLang/julia/issues/48457

### CODE REVIEW

**Comment by mikmoore** [comment]
This fix seems pretty limited. What about `inv(complex(0.2,1e-320))`? I assume this will still show the same issue on the real part. But fixing that is likely more involved and maybe should be a different PR.

File: base/complex.jl [hunk 1]
```diff
@@ -472,6 +472,11 @@ function inv(z::Complex{T}) where T<:Union{Float16,Float32}
 end
 function inv(w::ComplexF64)
     c, d = reim(w)
+    if d === 0.0
+        return ComplexF64(1/c, -0.0)
+    elseif d === -0.0
+        return ComplexF64(1/c,  0.0)
+    end
```
Comment (oscardssmith) [inline base/complex.jl:479]:
```suggestion
    if d == 0.0
        return ComplexF64(1/c, -d)
    end
```

File: base/complex.jl [hunk 2]
```diff
@@ -476,6 +476,8 @@ function inv(w::ComplexF64)
         return ComplexF64(1/c, -0.0)
     elseif d === -0.0
         return ComplexF64(1/c,  0.0)
+    if d == 0.0
+        return ComplexF64(1/c, -d)
```
Comment (oscardssmith) [inline base/complex.jl:480]: this suggestion didn't apply properly. This branch was supposed to replace the `if` and `elseif` above it.

**Comment by oscardssmith** [comment]
The slightly more general case is
```
if absd < sqrt(eps(Float64))*absc
    return complex(inv(c), -d/c)
end
```
This works because in this case `c^2+d^2 == c^2` so we can simplify.

**Comment by mikmoore** [comment]
Sorry for not taking a deeper look at this the first time. This doesn't seem like the right place for a fix. This change seems worse than doing nothing, since `inv(complex(0.2,0.0))` won't agree with `inv(complex(0.2,0.0)*im)*im`.

It appears that the fastpath `conj(w) / muladd(cd, cd, dc*dc)` computation is to blame here. For `complex(c,0.0)` this becomes `c/c^2 != 1/c`. It looks like the solution is simply to drop the fastpath. The slowpath `robust_cinv` will compute this correctly.

**Comment by KristofferC** [comment]
>  This change seems worse than doing nothing, since inv(complex(0.2,0.0)) won't agree with inv(complex(0.2,0.0)*im)*im

We can add a special case for zero real as well ;)

**Comment by oscardssmith** [comment]
IMO the fast path is worth it. It is occasionally slightly inaccurate, but it is 2x faster and is always pretty accurate.

**Comment by mikmoore** [comment]
Alright. I'm having trouble imagining other inputs where the fastpath result would be egregious. Just make sure that the extra checks aren't eroding all the benefit over a pure-slowpath version (don't need to post benchmarks, just know so).

**Comment by KristofferC** [comment]
Just to get a second data point I tried Rust (which is not really known for its numerics but still):

```rust
fn main() {
    println!("Complex float: {}", 1.0 / num::complex::Complex::new(0.2, 0.0));
    println!("Float: {}", 1.0 / 0.2);
}

// Printing:
Complex float: 4.999999999999999+0i
Float: 5
```

**Comment by mikmoore** [comment]
> ```rust
> 1.0 / num::complex::Complex::new(0.2, 0.0)
> ```

Does Rust have an `inv` operation? I'd check that, too. In any case, we're not alone in this. I checked MATLAB and it at least manages to give the `5.0` answer (even accounting for the dirty tricks they play with rounding printed numbers).

In Julia, I would accept `1.0/x` and `inv(x)` giving these different results (or at least `2.0/x` and `2.0*inv(x)`, were we to make a special case for `1.0`). The reason is that I would want `inv(complex(a,0)) == 1/a` while I would want `complex(a,b)/complex(a,b) == 1`. Satisfying these both with a unified function requires branching to different computations.

Another thought: our automatic differentiation libraries have explicit definitions of these computations, right? They don't just derive the result via these implementations? These sorts of special cases ruin derivatives.

**Comment by mikmoore** [comment]
I think the resulting derivatives would still be incorrect based on `complex(inv(c),-d/c)`. Neither of those terms includes the negligible-value parts that may have non-negligible derivatives. The errors would be small in many cases, but definitely present and almost certainly enough to cause problems.

**Comment by oscardssmith** [comment]
I've done the math (i.e. typed it into wolfram alpha) and the derivatives work out correct to within epsilon also.

**Comment by KristofferC** [comment]
> These sorts of special cases ruin derivatives. 

> I think the resulting derivatives would still be incorrect 

These are imo bugs in the AD libraries. For example, ForwardDiff handles these type of problems on its master branch (by not considering Dual numbers with non-zero dual part to be equal to a real number).

**Comment by mikmoore** [comment]
> I've done the math (i.e. typed it into wolfram alpha) and the derivatives work out correct to within epsilon also.

```julia-repl
julia> using ForwardDiff

julia> ForwardDiff.jacobian(((c,d),)->[c,-d]/(c^2+d^2),[0.2,0.0])
2×2 Matrix{Float64}:
 -25.0    0.0
   0.0  -25.0

julia> ForwardDiff.jacobian(((c,d),)->[inv(c),-d/c],[0.2,0.0])
2×2 Matrix{Float64}:
 -25.0  -0.0
   0.0  -5.0
```
Only the first version produces the correct result. Higher derivatives will be worse.

If `ForwardDiff` etc refutes equality with nonzero derivatives that's a start, but I can't imagine it does the same for inequalities like we're proposing for "small" `d`. And if it does refute equality then it will change the results of computations.

Sorry, this might be an unnecessary diversion. I'll let the issue go unless someone wants to engage further.

**Comment by oscardssmith** [comment]
yeah that's because I typod. It was supposed to be `inv(c)*Complex(1, d/c)`

**Comment by mikmoore** [comment]
Thanks. I should have caught that.

The point on higher derivatives stands
```julia-repl
julia> ForwardDiff.hessian(((c,d),)->inv(c),[0.2,0.0])
2×2 Matrix{Float64}:
 250.0  0.0
   0.0  0.0

julia> ForwardDiff.hessian(((c,d),)->c/(c^2+d^2),[0.2,0.0])
2×2 Matrix{Float64}:
 250.0     0.0
   0.0  -250.0
```

**Comment by oscardssmith** [comment]
Is this ready to merge?

**Comment by KristofferC** [comment]
I fixed the test upstream so I vote for not changing anything until someone else complains about it.

**Pull request closed without merging**